runtime/src/accounts_background_service.rs

```diff
@@ -169,23 +174,60 @@ impl SnapshotRequestHandler {
+                let snapshot_type = if block_height
+                    % self.snapshot_config.full_snapshot_archive_interval_slots
+                    == 0
+                {
+                    Some(SnapshotType::FullSnapshot)
+                } else if block_height
+                    % self
+                        .snapshot_config
+                        .incremental_snapshot_archive_interval_slots
+                    == 0
+                {
+                    Some(SnapshotType::IncrementalSnapshot(last_full_snapshot_slot))
+                } else {
+                    None
+                };
+
                 // Generate an accounts package
+                let mut attempts = 0;
                 let mut snapshot_time = Measure::start("snapshot_time");
-                let r = snapshot_utils::snapshot_bank(
-                    &snapshot_root_bank,
-                    status_cache_slot_deltas,
-                    &self.accounts_package_sender,
-                    &self.snapshot_config.bank_snapshots_dir,
-                    &self.snapshot_config.snapshot_archives_dir,
-                    self.snapshot_config.snapshot_version,
-                    &self.snapshot_config.archive_format,
-                    hash_for_testing,
-                );
-                if r.is_err() {
+                loop {
+                    attempts += 1;
+                    let result = snapshot_utils::snapshot_bank(
+                        &snapshot_root_bank,
+                        status_cache_slot_deltas.clone(),
+                        &self.accounts_package_sender,
+                        &self.snapshot_config.bank_snapshots_dir,
+                        &self.snapshot_config.snapshot_archives_dir,
+                        self.snapshot_config.snapshot_version,
+                        self.snapshot_config.archive_format,
+                        hash_for_testing,
+                        snapshot_type,
+                    );
+
+                    if result.is_ok() {
+                        break;
+                    }
+
                     warn!(
-                        "Error generating snapshot for bank: {}, err: {:?}",
+                        "Error generating snapshot for bank. attempts: {}, slot: {}, err: {:?}",
+                        attempts,
                         snapshot_root_bank.slot(),
-                        r
+                        result
+                    );
+
+                    // Only require success for full snapshots.  Otherwise just log the error and
+                    // continue on.
+                    if snapshot_type != Some(SnapshotType::FullSnapshot) {
+                        break;
+                    }
+
+                    assert!(
+                        attempts <= SNAPSHOT_BANK_MAX_ATTEMPTS,
+                        "Unable to snapshot bank intended for full snapshot!"
                     );
                 }
```

Comment:
This new loop is to retry full snapshots. They must succeed since we are calculating the last full snapshot slot instead of having SnapshotPackagerService tell us.

Notice that there's a `clone()` on the `status_cache_slot_deltas` now, which is a bummer. Maybe there's another way to do this?

Reply:
Yeah I think you would have to return the `status_cache_slot_deltas` in the error result of `snapshot_bank`. Maybe a `SnapshotBankArgs` that is the `snapshot_bank` arguments and is also that returned on error?

Reply:
Thanks, I'll give that a whirl!

Reply:
Now we aren't doing retries in ABS, so no more clone. Yay!